File: agent/src/main/java/com/codedx/codepulse/agent/agent/DefaultTraceAgent.java
/* Code Pulse: a real-time code coverage tool, for more information, see <http://code-pulse.com/>
 *
 * Copyright (C) 2014-2017 Code Dx, Inc. <https://codedx.com/>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.codedx.codepulse.agent.agent;

import java.io.File;
import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.Semaphore;

import com.codedx.codepulse.agent.TraceAgent;
import com.codedx.codepulse.agent.control.ConfigurationHandler;
import com.codedx.codepulse.agent.control.Controller;
import com.codedx.codepulse.agent.control.HeartbeatInformer;
import com.codedx.codepulse.agent.control.ModeChangeListener;
import com.codedx.codepulse.agent.control.StateManager;
import com.codedx.codepulse.agent.data.MessageDealerTraceDataCollector;
import com.codedx.codepulse.agent.errors.AgentErrorListener;
import com.codedx.codepulse.agent.errors.ErrorHandler;
import com.codedx.codepulse.agent.errors.MinlogListener;
import com.codedx.codepulse.agent.message.BufferService;
import com.codedx.codepulse.agent.message.MessageDealer;
import com.codedx.codepulse.agent.message.MessageSenderManager;
import com.codedx.codepulse.agent.message.PooledBufferService;
import com.codedx.codepulse.agent.protocol.ProtocolVersion;
import com.codedx.codepulse.agent.protocol.ProtocolVersion3;
import com.codedx.codepulse.agent.trace.TraceDataCollector;
import com.codedx.codepulse.agent.util.ShutdownHook;
import com.codedx.codepulse.agent.util.SocketFactory;
import com.codedx.codepulse.agent.common.config.RuntimeAgentConfigurationV1;
import com.codedx.codepulse.agent.common.config.StaticAgentConfiguration;
import com.codedx.codepulse.agent.common.connect.SocketConnection;
import com.codedx.codepulse.agent.common.message.AgentOperationMode;
import com.codedx.codepulse.agent.common.queue.BufferPool;

import com.codedx.bytefrog.instrumentation.id.*;
import com.codedx.bytefrog.util.Logger;
import com.esotericsoftware.minlog.Log;

/**
 * Concrete Agent implementation, manages the entire trace.
 *
 * @author RobertF
 */
public class DefaultTraceAgent implements TraceAgent
{
	private static long ConnectSleep = 1000;

	private SocketFactory socketFactory;
	private StaticAgentConfiguration staticConfig;
	private RuntimeAgentConfigurationV1 config;

	private final Semaphore startMutex = new Semaphore(0);
	private final ProtocolVersion protocol = new ProtocolVersion3();
	private MinlogListener logger = null;
	private ClassIdentifier classIdentifier = new ClassIdentifier();
	private MethodIdentifier methodIdentifier = new MethodIdentifier();
	private TraceDataCollector dataCollector;
	private StateManager stateManager;
	private Controller controller;
	private BufferPool bufferPool;
	private BufferService bufferService;
	private MessageDealer messageFactory;
	private MessageSenderManager senderManager;
	private boolean isStarted = false;
	private boolean isKilled = false;

	public ClassIdentifier getClassIdentifier() { return classIdentifier; }
	public MethodIdentifier getMethodIdentifier() { return methodIdentifier; }

	public DefaultTraceAgent(StaticAgentConfiguration staticConfig)
	{
		this.staticConfig = staticConfig;
		String error = "Failed to configure logger";

		try
		{
			// add a log listener now. Wait until initialization is done to add
			// the Agent listener
			String logFilename = staticConfig.getLogFilename();
			File logFile = logFilename != null ? new File(logFilename) : null;
			if (logFile != null) {
				Log.setLogger(new Logger(logFile));
				logger = new MinlogListener();
				ErrorHandler.addListener(logger);
			}

			error = "Failed to initialize state manager";

			stateManager = new StateManager();

			ModeChangeListener modeListener = new ModeChangeListener()
			{
				@Override
				public void onModeChange(AgentOperationMode oldMode, AgentOperationMode newMode)
				{
					if (oldMode == AgentOperationMode.Initializing
							&& newMode != AgentOperationMode.Shutdown)
						start();
					else if (newMode == AgentOperationMode.Shutdown)
						new Thread(new Runnable()
						{
							@Override
							public void run()
							{
								try
								{
									waitForSenderManager();
								}
								catch (InterruptedException ex)
								{
									Thread.currentThread().interrupt();
								}

								closeConnections();
							}
						}).start();
					else if (newMode == AgentOperationMode.Suspended)
						try
						{
							// alert HQ that there's a break in the data due to
							// the suspend
							controller.sendDataBreak(messageFactory.getCurrentSequence());
						}
						catch (IOException e)
						{
							ErrorHandler.handleError("Error sending data break message", e);
						}
				}
			};

			stateManager.addListener(modeListener);
			error = null;
		}
		catch (Exception e)
		{
			ErrorHandler.handleError("error initializing trace agent", e);
		}
		finally
		{
			if (error != null)
			{
				throw new RuntimeException("Agent Initialization Error: " + error);
			}
		}
	}

	@Override
	public boolean connect(int timeout) throws InterruptedException
	{
		String error = "";
		long timeoutExpire = System.currentTimeMillis() + timeout * 1000;

		try
		{
			error = "Failed to connect to HQ";
			socketFactory = new SocketFactory(staticConfig.getHqHost(), staticConfig.getHqPort());

			Socket controlSocket = null;

			while (controlSocket == null
					&& (timeout == 0 || System.currentTimeMillis() <= timeoutExpire))
			{
				try
				{
					controlSocket = socketFactory.connect();
				}
				catch (IOException e)
				{
					// try again...
					controlSocket = null;
					Thread.sleep(ConnectSleep);
				}
			}

			if (controlSocket == null)
			{
				// didn't connect within timeout period
				error = null;
				return false;
			}

			SocketConnection controlConnection = new SocketConnection(controlSocket, true, true);

			error = "Failed to get configuration from HQ";
			config = protocol.getControlConnectionHandshake().performHandshake(controlConnection);
			if (config == null)
			{
				controlSocket.close();
				ErrorHandler.handleError("did not receive valid configuration during handshake");
				return false;
			}

			HeartbeatInformer informer = new HeartbeatInformer()
			{
				@Override
				public int getSendQueueSize()
				{
					if (bufferPool != null)
						return bufferPool.numReadableBuffers();

					return 0;
				}

				@Override
				public AgentOperationMode getOperationMode()
				{
					return stateManager.getCurrentMode();
				}
			};

			ConfigurationHandler configHandler = new ConfigurationHandler()
			{
				@Override
				public void onConfig(RuntimeAgentConfigurationV1 newConfig)
				{
					// apply new configuration
					if (stateManager.getCurrentMode() != AgentOperationMode.Initializing)
					{
						ErrorHandler
								.handleError("reconfiguration is only valid while agent is initializing");
						return;
					}

					config = newConfig;
					controller.setHeartbeatInterval(config.getHeartbeatInterval());
				}
			};

			controller = new Controller(controlConnection, protocol, config.getHeartbeatInterval(),
					stateManager.getControlMessageHandler(), configHandler, informer);

			error = "Failed to start the agent controller";
			controller.start();
			error = null;

			// Now that everything is set up properly, it is safe to add an
			// AgentErrorListener.
			ErrorHandler.addListener(new AgentErrorListener(this));

			// successful connection
			return true;
		}
		catch (InterruptedException e)
		{
			// rethrow
			throw e;
		}
		catch (Exception e)
		{
			ErrorHandler.handleError("unhandled error connecting trace agent", e);
			return false;
		}
		finally
		{
			if (error != null)
			{
				throw new RuntimeException("Agent Connection Error: " + error);
			}
		}
	}

	@Override
	public void prepare()
	{
		initializeSender();
	}

	private void initializeSender()
	{
		try
		{
			// figure out the buffer count and sizes for the BufferPool
			int memBudget = config.getBufferMemoryBudget();
			int bufferLength = decideBufferLength(memBudget);
			int numBuffers = memBudget / bufferLength;

			// set up the queue/message factory
			bufferPool = new BufferPool(numBuffers, bufferLength);
			bufferService = new PooledBufferService(bufferPool, config.getQueueRetryCount());
			messageFactory = new MessageDealer(protocol.getMessageProtocol(), bufferService, classIdentifier, methodIdentifier);
			dataCollector = new MessageDealerTraceDataCollector(messageFactory);

			senderManager = new MessageSenderManager(socketFactory,
					protocol.getDataConnectionHandshake(), bufferPool, config.getNumDataSenders(),
					config.getRunId());
			senderManager.start();

			stateManager.addListener(bufferService.getModeChangeListener());
		}
		catch (Exception e)
		{
			ErrorHandler.handleError("error initializing trace agent and connecting to HQ", e);

			throw new RuntimeException(
					"Agent Initialization Error: Failed to set up message sending system");
		}
	}

	private int decideBufferLength(int memBudget)
	{
		int len = 8192;
		int minBuffers = 10;
		while (memBudget / len < minBuffers)
		{
			if (len < 128)
			{
				throw new IllegalArgumentException(
						"Agent's memory budget is too low to accomodate enough sending buffer space");
			}
			len /= 2;
		}
		return len;
	}

	@Override
	public StaticAgentConfiguration getStaticConfig()
	{
		return staticConfig;
	}

	@Override
	public RuntimeAgentConfigurationV1 getConfig()
	{
		return config;
	}

	@Override
	public TraceDataCollector getDataCollector()
	{
		return dataCollector;
	}

	@Override
	public StateManager getStateManager()
	{
		return stateManager;
	}

	@Override
	public Controller getControlController()
	{
		return controller;
	}

	@Override
	public MessageSenderManager getSenderManager()
	{
		return senderManager;
	}

	private void registerShutdownHook()
	{
		ShutdownHook hook = new ShutdownHook()
		{
			@Override
			protected void onShutdown()
			{
				try
				{
					shutdownAndWait();
				}
				catch (InterruptedException e)
				{
					if (logger != null)
					{
						logger.onErrorReported("interrupted while waiting for final shutdown", null);
					}
				}

				closeConnections();

				if (logger != null)
				{
					logger.close();
				}
			}
		};

		hook.registerHook();
	}

	@Override
	public void start()
	{
		isStarted = true;
		startMutex.release();
	}

	@Override
	public void killTrace(String errorMessage)
	{
		if (!isKilled)
		{
			// send error (if we can)
			try
			{
				controller.sendError(errorMessage);
			}
			catch (IOException e)
			{
				// log directly
				if (logger != null)
				{
					logger.onErrorReported(
							String.format("IO exception sending error message: %s", e.getMessage()),
							null);
				}
			}

			// then force a suspend and unpause
			if (bufferService != null)
			{
				// by forcing a suspend and unpause, the code being traced
				// should be able to keep running.
				bufferService.setSuspended(true);
				bufferService.setPaused(false);
			}

			shutdown();

			// if we haven't been started yet, we can exit
			if (!isStarted)
				System.exit(Integer.MIN_VALUE);

			isKilled = true;
		}
	}

	@Override
	public void shutdown()
	{
		stateManager.triggerShutdown();
	}

	@Override
	public void shutdownAndWait() throws InterruptedException
	{
		shutdown();
		waitForSenderManager();
	}

	public void closeConnections()
	{
		senderManager.shutdown();
		controller.shutdown();
	}

	private void waitForSenderManager() throws InterruptedException
	{
		int sleepInterval = config.getHeartbeatInterval();

		// poll send queue until it empties out
		do
		{
			try
			{
				Thread.sleep(sleepInterval);
			}
			catch (InterruptedException e)
			{
				ErrorHandler.handleError("interrupted while waiting for send queue to empty", e);
			}
		}
		while (!isKilled && (!senderManager.isIdle() || bufferPool.numReadableBuffers() > 0));
	}

	@Override
	public void waitForStart() throws InterruptedException
	{
		// wait for signal to start
		startMutex.acquire();

		// now that we're ready to start, register the shutdown hook
		registerShutdownHook();
	}
}

File: dotnet-tracer/main/CodePulse.Client/Agent/DefaultTraceAgent.cs
// Copyright 2017 Secure Decisions, a division of Applied Visions, Inc. 
// Permission is hereby granted, free of charge, to any person obtaining a copy of 
// this software and associated documentation files (the "Software"), to deal in the 
// Software without restriction, including without limitation the rights to use, copy, 
// modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
// and to permit persons to whom the Software is furnished to do so, subject to the 
// following conditions:
// 
// The above copyright notice and this permission notice shall be included in all copies 
// or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION 
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// This material is based on research sponsored by the Department of Homeland
// Security (DHS) Science and Technology Directorate, Cyber Security Division
// (DHS S&T/CSD) via contract number HHSP233201600058C.

using System;
using System.Threading;
using System.Threading.Tasks;
using CodePulse.Client.Config;
using CodePulse.Client.Connect;
using CodePulse.Client.Control;
using CodePulse.Client.Data;
using CodePulse.Client.Errors;
using CodePulse.Client.Init;
using CodePulse.Client.Instrumentation.Id;
using CodePulse.Client.Message;
using CodePulse.Client.Protocol;
using CodePulse.Client.Queue;
using CodePulse.Client.Trace;
using CodePulse.Client.Util;

namespace CodePulse.Client.Agent
{
    public class DefaultTraceAgent : ITraceAgent
    {
        private SocketFactory _socketFactory;

        private readonly IMessageProtocol _messageProtocol;
        private readonly IProtocolVersion _protocolVersion;

        private readonly StateManager _stateManager;

        private volatile bool _isKilled;
        private readonly object _isKilledLock = new object();

        private volatile bool _isShutdown;
        private readonly object _isShutdownLock = new object();

        private BufferPool _bufferPool;
        private BufferService _bufferService;
        private Controller _controller;

        private readonly HeartbeatInformer _heartbeatInformer;
        private readonly ConfigurationHandler _configurationHandler;

        private readonly ManualResetEvent _startEvent = new ManualResetEvent(false);
        private readonly ManualResetEvent _shutdownEvent = new ManualResetEvent(false);

        private readonly IErrorHandler _errorHandler = new ErrorHandler();

        private MessageSenderManager _messageSenderManager;

        public StaticAgentConfiguration StaticAgentConfiguration { get; }

        public RuntimeAgentConfiguration RuntimeAgentConfiguration { get; private set; }

        public ClassIdentifier ClassIdentifier { get; } = new ClassIdentifier();
        public MethodIdentifier MethodIdentifier { get; } = new MethodIdentifier();

        public ITraceDataCollector TraceDataCollector { get; private set; }

        public bool IsKilled
        {
            get => _isKilled;
            set => _isKilled = value;
        }

        public bool IsShutdown
        {
            get => _isShutdown;
            set => _isShutdown = value;
        }

        public DefaultTraceAgent(StaticAgentConfiguration staticAgentConfiguration)
        {
            StaticAgentConfiguration = staticAgentConfiguration ?? throw new ArgumentNullException(nameof(staticAgentConfiguration));

            _protocolVersion = new ProtocolVersion(_errorHandler);
            _messageProtocol = _protocolVersion.MessageProtocol;

            _heartbeatInformer = new HeartbeatInformer(this);
            _configurationHandler = new ConfigurationHandler(this);

            _stateManager = new StateManager(_errorHandler);
            _stateManager.AddListener(new ModeChangeListener(this));

            _errorHandler.ErrorOccurred += (sender, args) =>
            {
                var exceptionMessage = "n/a";
                if (args.Item2 != null)
                {
                    exceptionMessage = args.Item2.Message;
                }
                StaticAgentConfiguration.Logger.Error($"{args.Item1}: {exceptionMessage}", args.Item2);
            };
        }

        public bool Connect()
        {
            SocketConnection socketConnection;
            try
            {
                _socketFactory = new SocketFactory(StaticAgentConfiguration.HqHost,
                    StaticAgentConfiguration.HqPort,
                    StaticAgentConfiguration.ConnectTimeout);

                var socket = _socketFactory.Connect();
                if (socket == null)
                {
                    return false;
                }

                socketConnection = new SocketConnection(socket);
            }
            catch (Exception ex)
            {
                _errorHandler.HandleError("Failed to connect to HQ.", ex);
                return false;
            }

            try
            {
                try
                {
                    RuntimeAgentConfiguration = _protocolVersion.ControlConnectionHandshake.PerformHandshake(socketConnection, StaticAgentConfiguration.ProjectId);
                }
                catch (HandshakeException ex)
                {
                    socketConnection.Close();
                    _errorHandler.HandleError("Unable to perform control connection handshake.", ex);
                    return false;
                }
            }
            catch (Exception ex)
            {
                _errorHandler.HandleError("Failed to get configuration from HQ.", ex);
                return false;
            }

            _controller = new Controller(socketConnection,
                _protocolVersion,
                RuntimeAgentConfiguration.HeartbeatInterval,
                _stateManager.ControlMessageHandler,
                _configurationHandler,
                _heartbeatInformer,
                _errorHandler,
                StaticAgentConfiguration.Logger);

            _errorHandler.ErrorOccurred += (sender, args) => { RunKillTraceTask(args.Item1); };

            return true; 
        }

        public bool Prepare()
        {
            if (IsKilled)
            {
                return false;
            }

            try
            {
                var memBudget = RuntimeAgentConfiguration.BufferMemoryBudget;
                var bufferLength = DecideBufferLength(memBudget);
                var numBuffers = memBudget / bufferLength;

	            var logger = StaticAgentConfiguration.Logger;

				_bufferPool = new BufferPool(numBuffers, bufferLength, logger);
                _bufferService = new PooledBufferService(_bufferPool);
                TraceDataCollector = new TraceDataCollector(_messageProtocol, _bufferService, ClassIdentifier, MethodIdentifier, _errorHandler, logger);

                _messageSenderManager = new MessageSenderManager(_socketFactory,
                    _protocolVersion.DataConnectionHandshake,
                    _bufferPool,
                    RuntimeAgentConfiguration.NumDataSenders,
                    RuntimeAgentConfiguration.RunId,
                    _errorHandler,
                    logger);

                return true;
            }
            catch (Exception e)
            {
                _errorHandler.HandleError("Error initializing trace agent and connecting to HQ.", e);
                return false;
            }
        }

        public void Start()
        {
            _startEvent.Set();
        }

        public void WaitForStart()
        {
            _startEvent.WaitOne();
        }

        public void Shutdown()
        {
            _stateManager.TriggerShutdown();
        }

        public void WaitForShutdown()
        {
            _shutdownEvent.WaitOne();
        }

        private void KillTrace(string errorMessage)
        {
            var lockAcquired = Monitor.TryEnter(_isKilledLock);
            if (!lockAcquired)
            {
                return;
            }

            try
            {
                if (IsKilled)
                {
                    return;
                }

                try
                {
                    _controller.SendError(errorMessage);
                }
                catch (Exception ex)
                {
                    _errorHandler.HandleError("Exception occurred while sending error message.", ex);
                }

                Shutdown();

                // release any callers awaiting a start that will not occur
                _startEvent.Set();

                IsKilled = true;
            }
            finally
            {
                Monitor.Exit(_isKilledLock);
            }
        }

        private void ShutdownAgent()
        {
            var lockAcquired = Monitor.TryEnter(_isShutdownLock);
            if (!lockAcquired)
            {
                return;
            }

            try
            {
                if (IsShutdown)
                {
                    return;
                }

                // the TraceDataCollector shutdown will use the buffer service to send any pending 
                // method-entry messages - if the service is suspended, the messages will not reach Code Pulse
                TraceDataCollector?.Shutdown();

                if (_bufferService != null)
                { 
                    _bufferService.SetSuspended(true);
                    if (_bufferService.IsPaused)
                    {
                        _bufferService.SetPaused(false);
                    }
                }

                WaitForSenderManager();
                _messageSenderManager?.Shutdown();
                _controller?.Shutdown();

                IsShutdown = true;

                _shutdownEvent.Set();
            }
            finally
            {
                Monitor.Exit(_isShutdownLock);
            }
        }

        private void RunShutdownAgentTask()
        {
            Task.Run(() =>
            {
                ShutdownAgent();
            });
        }

        private void RunKillTraceTask(string errorMessage)
        {
            Task.Run(() =>
            {
                KillTrace(errorMessage);
            });
        }

        private void WaitForSenderManager()
        {
            if (_messageSenderManager == null)
            {
                return;
            }

            var sleepInterval = RuntimeAgentConfiguration.HeartbeatInterval;
            do
            {
                Thread.Sleep(sleepInterval);
            }
            while (!IsKilled && (!_messageSenderManager.IsIdle || _bufferPool.ReadableBuffers > 0));
        }

        private int DecideBufferLength(int memBudget)
        {
            var len = 8192;
            var minBuffers = 10;
            while (memBudget / len < minBuffers)
            {
                if (len < 128)
                {
                    throw new ArgumentException("Agent's memory budget is too low to accomodate enough sending buffer space");
                }
                len /= 2;
            }
            return len;
        }

        private class HeartbeatInformer : IHeartbeatInformer
        {
            private readonly DefaultTraceAgent _agent;

            public AgentOperationMode OperationMode => _agent._stateManager.CurrentMode;

            public int SendQueueSize => _agent._bufferPool?.ReadableBuffers ?? 0;

            public HeartbeatInformer(DefaultTraceAgent agent)
            {
                _agent = agent;
            }
        }

        private class ConfigurationHandler : IConfigurationHandler
        {
            private readonly DefaultTraceAgent _agent;
            
            public ConfigurationHandler(DefaultTraceAgent agent)
            {
                _agent = agent;
            }

            public void OnConfig(RuntimeAgentConfiguration config)
            {
                if (_agent._stateManager.CurrentMode != AgentOperationMode.Initializing)
                {
                    throw new InvalidOperationException("Reconfiguration is only valid while agent is running");
                }

                _agent.RuntimeAgentConfiguration = config;
                _agent._controller.SetHeartbeatInterval(config.HeartbeatInterval);
            }
        }

        private class ModeChangeListener : IModeChangeListener
        {
            private readonly DefaultTraceAgent _agent;

            public ModeChangeListener(DefaultTraceAgent agent)
            {
                _agent = agent;
            }

            public void OnModeChange(AgentOperationMode oldMode, AgentOperationMode newMode)
            {
                if (oldMode == AgentOperationMode.Initializing && newMode != AgentOperationMode.Shutdown)
                {
                    _agent.Start();
                    return;
                }

                switch (newMode)
                {
                    case AgentOperationMode.Initializing:
                        break;

                    case AgentOperationMode.Tracing:
                        switch (oldMode)
                        {
                            case AgentOperationMode.Paused:
                                _agent._bufferService?.SetPaused(false);
                                break;
                            case AgentOperationMode.Suspended:
                                _agent._bufferService?.SetSuspended(false);
                                break;
                        }
                        break;

                    case AgentOperationMode.Paused:
                        _agent._bufferService?.SetPaused(true);
                        break;

                    case AgentOperationMode.Suspended:
                        try
                        {
                            _agent._controller.SendDataBreak(_agent.TraceDataCollector.SequenceId);
                        }
                        catch (Exception ex)
                        {
                            _agent._errorHandler.HandleError("Error sending data break message.", ex);
                        }
                        _agent._bufferService?.SetSuspended(true);
                        break;

                    case AgentOperationMode.Shutdown:
                        _agent.RunShutdownAgentTask();
                        break;

                    default:
                        throw new ArgumentOutOfRangeException(nameof(newMode), newMode, null);
                }
            }
        }
    }
}

